### DISCUSSION TITLE
odd behavior with tab panel

### DISCUSSION THREAD

**Discussion by vickorian** [zauberzeug/nicegui#3194] [Q&A]
### Question

I have an odd behavior in a tab panel that's a bit frustrating. When adding rows, they're placed side by side, great. if i set the rows width to "w-2/3" and "w-1/3" they'll be placed side by side as long as the containing row contains the ".classes('m-0 p-0 no-wrap")". Fine. The issue is that the panel won't be fully spanned by these and only will span the full panel if I pass a width in style with ".style("width: 800px") for the element in the "w-2/3" wrapper. If i don't pass style directly, it won't fully span the panel.

Even more frustrating, If I attempt the same setup and replace ui.editor or another element with a ui.aggrid, then the aggrid will not show at all and will only be visible if pass the "style()" method to the aggrid. Any advice for how to fix this would be great. 

Thanks.

**Comment by falkoschindler** [comment]
Hi @vickorian, I think it would help to see some code reproducing the problem.

**Reply by vickorian** [reply]
@falkoschindler Hey, i'll type this up in a bit more detail when i get a minute. I'm still trying to finish a project and can work around the issue as of now, but I think it might fall more into a bug/design problem. I"ll come up with some toy examples when i get some time to explore the problem some more.

**Reply by vickorian** [reply]
@falkoschindler Here's an interesting example that's easy to reproduce i just came across.

```py
with ui.column().classes('w-full items-center'):
    with ui.tabs() as tabs:
        ui.tab('CR', label='Create New Customer')
        ui.tab('UP', label='Update Existing Customer')
        ui.tab('CRC', label="Create New Case")
        ui.tab('UPC', label="Update existing case")
        ui.tab('FIN', label='Finances')

with ui.tab_panels(tabs, value=app.storage.user['tab']).classes('w-full').bind_value_to(app.storage.user, 'tab').on_value_change(vis_change):
    with ui.tab_panel('CR'):
        pass
    with ui.tab_panel('UP'):
        pass
    with ui.tab_panel('CRC'):
        pass
    with ui.tab_panel('UPC'):
        pass
    with ui.tab_panel('FIN'):
        with ui.row().classes('w-full no-wrap'):
            with ui.column().classes('w-1/6 h-full items-center border'):    
                with ui.tabs().props('vertical').classes('w-full') as tabs_FIN:
                    ui.tab("FINR", label="View Reports")
                    ui.tab("FINC", label="Create Payment Schedule")
                    ui.tab("FINP", label="Add Payment")      
                with ui.row().style('height: 500px'):
                    ui.space()
                ui.space()
                with ui.row().classes('items-center'):
                    ui.button('Testing button', color='violet')
    
            with ui.tab_panels(tabs_FIN, value="FINP").bind_value(app.storage.user, 'tab2'):
                with ui.tab_panel("FINP").classes('w-full items-center p-0 m-0'):
                    with ui.column().classes('w-full items-center p-0 m-0'):
                        with ui.row().classes('w-full p-0 m-0 border no-wrap'):
                            with ui.row().classes('w-1/3 p-0 m-0'):
                                ui.input(label="Case No.").style("width: 180px")
                                
                                
                            
                            with ui.row().classes('w-1/3 p-0 m-0'):
                                ui.select(["Supreme","Court of Appeals","Superior", "Business","State","Juvenile",
                                        "Probate","Magistrate","Civil","Municipal","Recorder's"], 
                                        label="Select Court").style("width: 180px")
                                
                                
                                
                            with ui.row().classes('w-1/3 p-0 m-0'):
                                ui.input(label="Address of Court").style("width: 180px")
                                ui.input(label="Zip of Court").style("width: 180px").props("mask=#####")
                                ui.input(label="Name of Judge").style("width: 180px")
                                ui.input(label="Date of Next appearance").style("width: 180px").props("type=date filled")
                                ui.input(label="Time of Next appearance").style("width: 180px").props("type=time filled")
```

Now this behaves properly. It spans the page and all the data inputs are where you'd expect them to be. But if you truncate the last row to only have one input instead of 5, the rows will all contract to be right next to each other and lack the expected separation.

**Reply by falkoschindler** [reply]
Why do you use `ui.row` instead of `ui.column`? A row with 5 inputs doesn't have enough space, so it occupies all it can get and wraps its content. When removing all but one input, it fits on one row and the `ui.row` doesn't need to grow.

**Reply by vickorian** [reply]
I tried all the variations between using columns, rows, combinations to get the formatting correct. In lots of cases, when i'd pass columns, it wouldn't size correctly or it would stack them vertically or it wouldn't fill the space. I framed out lots of these types of things using the "border" command in classes to visualize how each element would fit together and this orientation gave the "least" amount of kickback when styling at this depth.

When passing what you'd think would be the desired output in the format of "create a large column of rows, create rows within column, create columns within row for data elements" the formatting of the page never held up. Pieces would be placed randomly, elements would load on top of each other, spacing wouldn't be maintained.

**Reply by falkoschindler** [reply]
As far as I can tell, `ui.column` did work for me. But I'm not 100% sure I understand the problem. I suggest to further reduce the problem to the bare minimum and to solve it before further growing the UI. Small problems are easier to understand (citation needed) and also easier to communicate. In your case, I don't think you need two layers of tab panels to reproduce the layout issue.
